design/authenticators/authn_jwt/token_schema.md

```diff
@@ -0,0 +1,178 @@
+[//]: # "Elaborate on the issue you are writing a solution for"
+
+1. Required fields
+   1. In the authenticator policy, the user will have the ability to declare which claims must be in host annotations. As a result, also in the JWT token
+   2. When sending an authentication request, the authenticator MUST validate those claims - existence and content 
+2. Mapping 
+   1. The user should have the ability to map a given "technical" claim name to a more familiar / business one - i.e, in git-lab. Instead of mentioning in the policy "sub" claim, the user will have the ability to mention "job"
+
+## Solution
+User will define the token schema variables in the authenticator policy. The variables will be checked in the JWT authentication request.
+
+* The usage of this variables is optional and non mandatory. User can authenticate to JWT in the same way he authenticated in stage 1. Using the feature adds more security checks to the authentication.
+
+### User Interface
+
+#### Authenticator Policy
+
+Two new variables will be added to the authenticator policy:
+
+* mandatory_claims
+* claims_mapping
+
+```yaml
+    - !policy
+    id: conjur/authn-jwt/VendorY
+    body:
+    - !webservice
+    
+    - !variable
+      id: mandatory_claims
+      
+    - !variable
+      id: claims_mapping
+```
+
+#### Variable Values Example
+
+| Name             | Value                   | Description                                      |
+| ---------------- | ----------------------- | ------------------------------------------------ |
+| mandatory_claims | branch, job, aud        | List of claims as they appear in host annotation |
+| claims_mapping   | {ref: branch, sub: job} | Mapping between host annotation to JWT claim     |
+
+* For any claim mapped, we will add the host annotation to the `mandatory_claims` variable and not the JWT claim name.
+* The values above are examples and they can be any claims.
+
+
+## Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+
+The following class diagram, represent the flow of checking the mandatory claims and the claim mapping.
+
+### Class Diagrams
+
+
+![image-20210711171238701](token-schema-class-diagram.png)
+
+### Flow Explanation
+
+When validate_restrictions is called in JWTVendorConfiguration class the following will happen as a result:
+
+1. The `valitate_restrictions` function will call `CreateConstraintsFromPolicy` that create constraints object checking the hosts annotations.
+   1. It will create instance of `NonEmptyConstraint` to check we have at least one host annotation.
+   2. It will call `LoadMAndatoryClaims` command class to load the list of mandatory claims for the `mandatory_claims` variable
+   3. It will create `RequiredContraint` object from it
+2. The `validate_restrictions` functon will call `CreateValidatorFromPolicy` to create validator with the claims mapping loaded
```

Comment:
function*

Reply:
Why do we need this class `CreateValidatorFromPolicy` what her responsibility ?

You could directly create the `ValidateRestrictionOneToOne` which has a dependency of key mapping.
And maybe in the future when we will extend this functionality to more possibilities for example  `ValidateRestrictionWithConditions` or `ValidateRestrictionByRegex`
So you will need this extra layer of `CreateValidatorFromPolicy` to decide which object to create 

WDYT?